src/frontend/src/optimizer/mod.rs

```diff
@@ -849,15 +849,35 @@ impl PlanRoot {
         }
 
         let conflict_behavior = match on_conflict {
-            Some(on_conflict) => match on_conflict {
-                OnConflict::UpdateFull => ConflictBehavior::Overwrite,
-                OnConflict::Nothing => ConflictBehavior::IgnoreConflict,
-                OnConflict::UpdateIfNotNull => ConflictBehavior::DoUpdateIfNotNull,
-            },
-            None => match append_only {
-                true => ConflictBehavior::NoCheck,
-                false => ConflictBehavior::Overwrite,
-            },
+            EitherOnConflict::Ast(on_conflict) => {
+                if append_only {
+                    if row_id_index.is_some() {
+                        // Primary key will be generated, no conflict check needed.
+                        ConflictBehavior::NoCheck
+                    } else {
+                        // User defined PK on append-only table, enforce `DO NOTHING`.
+                        if let Some(on_conflict) = on_conflict
+                            && on_conflict != OnConflict::Nothing
+                        {
+                            return Err(ErrorCode::InvalidInputSyntax(
+                                "When PRIMARY KEY constraint applied to an APPEND ONLY table, \
+                                 the ON CONFLICT behavior must be DO NOTHING."
+                                    .to_owned(),
+                            )
+                            .into());
+                        }
+                        ConflictBehavior::IgnoreConflict
+                    }
+                } else {
+                    // Default to `UPDATE FULL` for non-append-only tables.
+                    match on_conflict.unwrap_or(OnConflict::UpdateFull) {
+                        OnConflict::UpdateFull => ConflictBehavior::Overwrite,
+                        OnConflict::Nothing => ConflictBehavior::IgnoreConflict,
+                        OnConflict::UpdateIfNotNull => ConflictBehavior::DoUpdateIfNotNull,
+                    }
+                }
+            }
+            EitherOnConflict::Resolved(b) => b,
```

Comment:
What about making this a method on `EitherOnConflict` to make the optimizer's code shorter?